RTD: add a "how to optimize a function" example

DougBurke opened the pull request:
# Summary

Add a section to the ReadTheDocs site explaining how to "optimize a function" as this is something we can do but only really with the low-level code.

# Details

This is an offshoot of our discussions with 3ML but it's something I wanted to know how to do. There are some holes in our documentation - in particular for the lmdif-needed least-squares return value - but it's a start. The question is whether it's a useful start.

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+********************
+Optimizing functions
```
I suggest "finding minima" or something like that. When I read it first, I thought this is about optimizing the code to make functions run faster. Maybe I'm too much of a coder an normal users won't think that way.

hamogu replied:
Maybe "minimizing functions"?

DougBurke replied:
I note that the scipy docs are "Optimization"

https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html

and we have a big section on "Optimisers" at https://sherpa.readthedocs.io/en/latest/optimisers/index.html so I'm less concerned about this issue than you are. I don't particularly like the title but I do not have the energy/time to come up with a better one.

I will add a link to the scipy docs to let users know what we're trying to do here.

hamogu replied:
OK, good point. Fine with me.

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+
+  def rosenbrock(x, y):
+      a, b = 2, 10
+      return (a - x) * (a - x) + b * (y - x * x) * (y - x * x)
```
```suggestion
      return (a - x)**2 + b * (y - x * x)**2
```
personal preference. I think it's easier to read because it's shorter, but either way is fine.

DougBurke replied:
How about

```suggestion
      return (a - x)**2  + b * (y - x**2)**2   # :-)
```

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+      stat = pa * pa + pb * pb
+      return stat, None
+
+The function can tehn be monimized by passing the routine, the
```
```suggestion
The function can then be minimized by passing the routine, the
```
I know you put that in jsut ot maek suer we pya attentions.

DougBurke replied:
Look. You thought we were learning about how to speed up code, not monomize functions!

hamogu replied:
monomized, sure. But monimized? https://in.pinterest.com/srsirisharao/monomize/

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+
+The function can tehn be monimized by passing the routine, the
+starting point, and the bounds, to the optimizer- in this case the
+:py:func:`~sherpa.optmethods.optfcts.minim` routine:
```
```suggestion
`~sherpa.optmethods.optfcts.minim` routine:
```

Should work here, too. But not 100% certain, need to check output at RTD if we go with this suggestion. (If it does not work, I need to tweak the conf setting more - not not for this release.)

DougBurke replied:
Interestingly they are **slightly** different, but not enough that I care about...

```
TEST1 `~sherpa.optmethods.optfcts.minim` routine

TEST2 :py:func:`~sherpa.optmethods.optfcts.minim` routine
```

creates

![image](https://user-images.githubusercontent.com/224638/133847219-fefa185c-7f6f-450d-bb24-4356fc570944.png)

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+  def to_optimize2(args):
+      x = args[0]
+      y = args[1]
+      pa = (a - x) * (a - b)
+      pb = b * (y - x * x) * (y - x * x)
```
```suggestion
      pa = (a - x)**2
      pb = b * (y - x * x)**2
```
Again, just my personal preference.

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+
+  def ngauss(x, ampl, pos, fwhm):
+      term = 4 * np.log(2)
+      numerator = ampl * np.exp(-term * (x - pos) * (x - pos) / (fwhm * fwhm))
```
Again `**2` to reduce the line length? This is long in my browser window, but of course it depends on the size of your screen and settings.

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+  >>> x = d[:, 0]
+  >>> y = d[:, 1]
+
+In this case the "callback" routine we want to give the minimizer a
+routine which will apply a least-squares minimizer to compare the
```
```suggestion
In this case the "callback" routine we want to give the minimizer is a
routine which will apply a least-squares minimizer to compare the
```

DougBurke replied:
I'm re-wording this entire single-sentence paragraph.

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+      return stat, None
+
+The function can tehn be monimized by passing the routine, the
+starting point, and the bounds, to the optimizer- in this case the
```
```suggestion
starting point, and the bounds, to the optimizer - in this case the
```

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+      return statval, delta
+
+.. note::
+   This function would normally be written as a closure to ensure
```
I don't know what a "closure" is, I would use a callable class that keep that info as a class attribute. But either way, the point is to point out that that's a simply example, so whatever works here.
Or maybe a callable class is just one example to implement a  "closure"? I need to google and learn that term!

DougBurke replied:
A closure "encloses" the symbols it uses and I have seen it used in python discussions - even Guido has used it ...

https://www.python.org/dev/peps/pep-3104/#id47

DougBurke replied:
I've rewritten this to say something "a class or a closure"

hamogu commented on docs/extra/optimize_function.rst:
```diff
@@ -0,0 +1,297 @@
+   that the use of global terms like `x`, `y`, `ngauss`, and `store` do not
+   cause problems.
+
+For the starting point and bounds I'm using to use a "guess" based on
```
```suggestion
For the starting point and bounds I'm using a "guess" based on
```

hamogu replied:
The rest of the test is written with "we do this". Now, you switch to "I". Intentional?

DougBurke replied:
I'm ambidextrous... We shoulf=d really scrub the documnetation for this sort of change in style - I do it a lot - but then I decide that consistency is a hobgoblin of minds that aren't knee deep in documentation and forget...

hamogu reviewed (changes requested):
The content it great, the examples are great. Very useful addition to the docs. I just have a few things related to typos or formatting.

anetasie commented:
@DougBurke Thanks for the examples. Playing with 3D plots is fun!
The new section looks good.

anetasie commented:
> @DougBurke Thanks for the examples. Playing with 3D plots is fun!
> The new section looks good.

btw. I note the same typo as @hamogu comment above.

dtnguyen2 reviewed (commented):
I know that this is about _how to optimize a function_
```
def cb(pars):
     model = ngauss(x, pars[0], pars[1], pars[2])
     delta = model - y
     statval = (delta * delta).sum()
     # Keep a record of the parameters we've visited
     store.add_row([statval] + list(pars))
     return statval, delta
```
but should we mention that a history of the parameters visited during a fit can be done with an extra argument to the ```fit()``` function with the optional ```outfile```?

DougBurke commented:
@dtnguyen2 - this is talking at a much lower part of the software stack than the outfile argument (ie there's 'fit' call here).

DougBurke commented:
I think I've addressed the review comments.

I am not convinced I'm doing a good job of describing what the callback routine should be returning - e.g. the least-squared difference in the first case is not well elucidated and the terms needed by lmdif is hand-waved over.

We do have https://docs.scipy.org/doc/scipy/reference/tutorial/optimize.html as a model
